File: py/aoc2024/day7.py
"""
Day 7:
"""

from typing import Callable, Iterable

SAMPLE_INPUT = """
190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20
"""


def _solve(data: str, op: Callable[[int, int], Iterable[int]]) -> int:
    total = 0
    for line in filter(None, data.splitlines()):
        lhs, rhs = line.split(":", maxsplit=1)
        lhs, rhs = int(lhs), [int(value) for value in rhs.split()]

        def ok(x, remaining):
            y = remaining[-1]
            return (
                x == y
                if len(remaining) == 1
                else any(ok(z, remaining[:-1]) for z in op(x, y))
            )

        if ok(lhs, rhs):
            total += lhs
    return total


def part1(data: str) -> int:
    """
    >>> part1(SAMPLE_INPUT)
    3749
    """

    def op(x, y):
        values = []
        if x >= y:
            values.append(x - y)
        if not x % y:
            values.append(x // y)
        return values

    return _solve(data, op)


def part2(data: str) -> int:
    """
    >>> part2(SAMPLE_INPUT)
    11387
    """

    def op(x, y):
        values = []
        if x >= y:
            values.append(x - y)
        if not x % y:
            values.append(x // y)
        x, y = str(x), str(y)
        if x.endswith(y) and len(x) > len(y):
            values.append(int(x[: -len(y)]))
        return values

    return _solve(data, op)


parts = (part1, part2)

File: hs/src/Day7.hs
{-# LANGUAGE OverloadedStrings #-}

-- |
-- Module:         Day7
-- Description:    <https://adventofcode.com/2024/day/7 Day 7: Bridge Repair>
module Day7 (part1, part2) where

import Common (readEntire, readSome)
import Control.Monad (ap)
import Control.Parallel.Strategies (parMap, rseq)
import Data.List.NonEmpty (NonEmpty ((:|)), nonEmpty)
import Data.List.NonEmpty qualified as NE (reverse)
import Data.Text (Text)
import Data.Text qualified as T (lines, stripPrefix)
import Data.Text.Read (Reader)
import Data.Text.Read qualified as T (decimal)

parseLine :: (Integral a) => Reader (a, NonEmpty a)
parseLine line = do
  (lhs, line') <- T.decimal line
  (rhs, line'') <- maybe (Left "") (readSome T.decimal) $ T.stripPrefix ": " line'
  pure ((lhs, NE.reverse rhs), line'')

parFilter :: (a -> Bool) -> [a] -> [a]
parFilter f = fmap fst . filter snd . (zip `ap` parMap rseq f)

solve :: (Integral a) => (a -> a -> [a]) -> Text -> Either String a
solve op input = sum . map fst . parFilter f <$> mapM (readEntire parseLine) (T.lines input)
  where
    f (x, y :| rest)
      | Just rest' <- nonEmpty rest = or [f (z, rest') | z <- x `op` y]
      | otherwise = x == y

op1, op2 :: (Integral a) => a -> a -> [a]
op1 x y = [x - y | x >= y] <> case x `divMod` y of (z, 0) -> [z]; _ -> []
op2 x y =
  op1 x y
    <> [ z
       | d <- take 1 $ dropWhile (<= y) $ iterate (* 10) 10,
         let (z, r) = x `divMod` d,
         z /= 0 && r == y
       ]

part1, part2 :: Text -> Either String Int
part1 = solve op1
part2 = solve op2
